src/common/libutil/fileref.c

```diff
@@ -0,0 +1,356 @@
+                goto error;
+            if (blobsize > cursor - offset)
+                blobsize = cursor - offset;
+            if (blobref_hash (hashtype,
+                              mapbuf + offset,
+                              blobsize,
+                              blobref,
+                              sizeof (blobref)) < 0)
+                goto error;
+            if (!(o = json_pack ("[I,I,s]", offset, blobsize, blobref))
+                || json_array_append_new (blobvec, o) < 0) {
+                json_decref (o);
+                goto nomem;
+            }
+            offset += blobsize;
+        }
+    }
+    return 0;
+nomem:
+    errno = ENOMEM;
+error:
+    return -1;
+}
+
+/* Optionally set fileref.data (string).
+ * If 'data' is NULL do nothing.
+ * If encode_base64 is false, assume 'data' is NULL terminated and use directly.
+ * If encode_base64 is true, base64-encode it first.
+ */
+static int set_optional_data (json_t *fileref,
+                              const void *data,
+                              size_t data_size,
+                              bool encode_base64)
+{
+    if (data) {
+        json_t *o;
+
+        if (encode_base64) {
+            char *buf = NULL;
+            size_t bufsize = base64_encoded_length (data_size) + 1; // +1 NULL
+
+            if (!(buf = malloc (bufsize)))
+                return -1;
+            if (base64_encode (buf, bufsize, data, data_size) < 0) {
+                free (buf);
+                errno = EINVAL;
+                return -1;
+            }
+            o = json_string (buf);
+            free (buf);
+        }
+        else
+            o = json_string (data);
+        if (!o
+            || json_object_set_new (fileref, "data", o) < 0) {
+            json_decref (o);
+            errno = ENOMEM;
+            return -1;
+        }
+    }
+    return 0;
+}
+
+/* Create fileref object representing regular file, symbolic link, or
+ * directory. 'fullpath' is used to open/stat the file, while 'path' is
+ * recorded in the fileref object.
+ */
+json_t *fileref_create_ex (const char *path,
+                           const char *fullpath,
+                           const char *hashtype,
+                           int chunksize,
+                           void **mapbufp,
+                           size_t *mapsizep,
+                           flux_error_t *error)
+{
+    json_t *blobvec = NULL;
+    json_t *o;
+    int fd = -1;
+    struct stat sb;
+    void *mapbuf = MAP_FAILED;
+    size_t mapsize = 0;
+    int saved_errno;
+    void *data = NULL;
+    ssize_t data_size = 0;
+    bool data_encode_base64 = false;
+    int rc;
+
+    if (chunksize < 0) {
+        errprintf (error, "chunksize cannot be negative");
+        goto inval;
+    }
+    if (!fullpath)
+        fullpath = path;
+    /* Refuse to put absolute paths in a fileref object.  Strip leading '/'.
+     */
+    while (*path == '/')
+        path++;
```

Comment:
Correct - it would be a bit too easy to do something like
```
$ flux filemap /g/gapps/mydata
$ flux mini run -o copyin ...
```
and then parallel overwrite the data in /usr/gapps.  I'll expand that comment a bit!